modules/javafx.graphics/src/main/java/com/sun/scenario/animation/LinearInterpolator.java

```diff
@@ -0,0 +1,211 @@
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.scenario.animation;
+
+import javafx.animation.Interpolator;
+import javafx.geometry.Point2D;
+import java.util.Arrays;
+
+/**
+ * Implementation of a piecewise linear interpolator as described by
+ * <a href="https://www.w3.org/TR/css-easing-2/#the-linear-easing-function">CSS Easing Functions Level 2</a>
+ */
+public final class LinearInterpolator extends Interpolator {
+
+    // Control points stored as [x0, y0, x1, y1, ...]
+    private final double[] controlPoints;
+
+    public LinearInterpolator(Point2D[] controlPoints) {
+        if (controlPoints == null) {
+            throw new NullPointerException("controlPoints cannot be null");
+        }
+
+        if (controlPoints.length == 0) {
+            throw new IllegalArgumentException("controlPoints cannot be empty");
+        }
+
+        int n = controlPoints.length;
+        this.controlPoints = new double[n * 2];
+
+        for (int i = 0; i < n; i++) {
+            Point2D p = controlPoints[i];
+            this.controlPoints[2 * i] = p.getX();
+            this.controlPoints[2 * i + 1] = p.getY();
+        }
+
+        canonicalize(this.controlPoints, n);
+    }
+
+    private static void canonicalize(double[] controlPoints, int n) {
+        // If the first control point has no input progress value, set it to 0.
+        if (Double.isNaN(controlPoints[0])) {
+            controlPoints[0] = 0.0;
+        }
+
+        // If the last control point has no input progress value, set it to 1.
+        int lastIdx = 2 * (n - 1);
+        if (Double.isNaN(controlPoints[lastIdx])) {
+            controlPoints[lastIdx] = 1.0;
+        }
+
+        // Ensure that the input progress value of each control point is greater than or equal to the
+        // input progress values of all preceding control points (monotonically non-decreasing).
+        double largestX = controlPoints[0];
```

Comment:
Question: what happens when a esquence is specified which is somehow invalid?  Will it throw an exception, write to stderr, or silently ignore?

(I can't think of an invalid sequence, maybe `0, 0 0% 0%, 0 -10%, NaN` ?)

Reply:
The sequence will always be successfully canonicalized, so there's never a log output or exception. In your example, the canonicalized easing function will be 0 for t=0, and NaN for t>0. If you use this function in an animation, you may get weird values and undefined things may happen. This is not unique to easing functions: we almost never handle NaNs in JavaFX. For example, you can construct a `Color` with NaN components (which is what would happen if you used your easing function in a color transition).